### PR TITLE
feat(shared): add vim.is_truthy()

### PR DESCRIPTION
**Pull request by uga-rosa**
When using `vim.fn`, there is a problem that vim script and lua handle truthy/falsy values differently.
`vim.is_truthy()` returns a boolean indicating whether a value is truthy/falsy in vim script.

### CODE REVIEW

**Comment by lewis6991** [comment]
Not really a fan of this. Longer term `vim.fn` will be phased out as the API becomes more complete.

The normal convention atm is just `vim.fn.foo(...) == 0`.

**Comment by uga-rosa** [comment]
Unless vim script itself is deprecated, it is still available to be manipulated from lua. For example, it is quite possible that a plugin written in vim script can be manipulated from lua. And there is no guarantee that plugin authors will only use 0 as a falsy value.

**Comment by clason** [comment]
There _is_ a convention in Vimscript, though. And if you use plugin functions, you are expected to read the documentation for it. (There _is_ documentation, right?!)

I agree that this specific function adds too much overhead for the relatively small convenience it offers.

**Comment by uga-rosa** [comment]
I'm mainly envisioning using this to enclose vim.fn when using it for conditional expressions such as if, but is this a case where users should write their own utilities to handle this?
```lua
-- A common bug
if vim.fn.filereadble('/path/to/file') then
  -- do something
end
```

**Comment by clason** [comment]
This is just 
```lua
-- A common bug
if vim.fn.filereadble('/path/to/file') == 1 then
  -- do something
end
```

I think it would be better to document this in `:h lua-guide`. Vimscript is messy, and trying to paper over that will just lead to even harder to diagnose bugs.

**Comment by zeertzjq** [comment]
Related Vim script function: `empty()`

**Comment by lewis6991** [comment]
Introducing `vim.is_truthy` won't prevent or even reduce that bug from happening.

> but is this a case where users should write their own utilities to handle this?

No utilities are needed, the correct way to write this is:

```lua
if vim.fn.filereadble('/path/to/file') == 1 then
  -- do something
end
```

**Comment by uga-rosa** [comment]
My actual use case is this
```lua
vim.bool_fn = setmetatable({}, {
  __index = function(_, key)
    return function(...)
      local v = vim.fn[key](...)
      return vim.is_truthy(v)
    end
  end,
})
```
```lua
if vim.bool_fn.filereadble('/path/to/file') then
  -- do something
end
```

**Comment by lewis6991** [comment]
Why not do:

```lua
vim.bool_fn = setmetatable({}, {
  __index = function(_, key)
    return function(...)
      local v = vim.fn[key](...)
      return not vim.fn.empty(v)
    end
  end,
})
```

File: runtime/doc/lua.txt [hunk 1]
```diff
@@ -1718,6 +1718,15 @@ is_callable({f})                                           *vim.is_callable()*
     Return: ~
         (boolean) `true` if `f` is callable, else `false`
 
+is_truthy({v})                                               *vim.is_truthy()*
+    Determine truthy/falsy in vim script.
```
Comment (clason) [inline runtime/doc/lua.txt:1722]: To be clear, what I am extremely hesitant about is to add such a single-purpose convenience function to the global `vim` module. The name even suggests a very generic (Lua) function, which will add to confusion.

**Comment by uga-rosa** [comment]
Aside from whether it should be mentioned in `:h lua-guide`, this PR is closed. Thanks for discussing!

**Pull request closed without merging**